test/integration/framework/etcd.go

```diff
@@ -17,93 +17,126 @@ limitations under the License.
-		if err != nil {
-			glog.Fatalf("Unable to make temp etcd data dir: %v", err)
-		}
-		glog.Infof("storing etcd data in: %v", etcdDataDir)
-
-		etcdCmd := exec.Command(
-			etcdPath,
-			"--data-dir",
-			etcdDataDir,
-			"--listen-client-urls",
-			GetEtcdURL(),
-			"--advertise-client-urls",
-			GetEtcdURL(),
-			"--listen-peer-urls",
-			"http://127.0.0.1:0",
-		)
-
-		stdout, err := etcdCmd.StdoutPipe()
-		if err != nil {
-			glog.Fatalf("Failed to run etcd: %v", err)
-		}
-		stderr, err := etcdCmd.StderrPipe()
+`
+
+// getEtcdPath returns a path to an etcd executable.
+func getEtcdPath() (string, error) {
+	bazelPath := filepath.Join(os.Getenv("RUNFILES_DIR"), "com_coreos_etcd/etcd")
+	p, err := exec.LookPath(bazelPath)
+	if err == nil {
+		return p, nil
+	}
+	return exec.LookPath("etcd")
+}
+
+// getAvailablePort returns a TCP port that is available for binding.
+func getAvailablePort() (int, error) {
+	l, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return 0, fmt.Errorf("could not bind to a port: %v", err)
+	}
+	// It is possible but unlikely that someone else will bind this port before we
+	// get a chance to use it.
+	defer l.Close()
+	return l.Addr().(*net.TCPAddr).Port, nil
+}
+
+// startEtcd executes an etcd instance. The returned function will signal the
+// etcd process and wait for it to exit.
+func startEtcd() (func(), error) {
+	etcdURL = env.GetEnvAsStringOrFallback("KUBE_INTEGRATION_ETCD_URL", "http://127.0.0.1:2379")
+	conn, err := net.Dial("tcp", strings.TrimPrefix(etcdURL, "http://"))
+	if err == nil {
+		glog.Infof("etcd already running at %s", etcdURL)
+		conn.Close()
+		return func() {}, nil
+	}
+	glog.V(1).Infof("could not connect to etcd: %v", err)
+
+	// TODO: Check for valid etcd version.
+	etcdPath, err := getEtcdPath()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, installEtcd)
+		return nil, fmt.Errorf("could not find etcd in PATH: %v", err)
+	}
+	etcdPort, err := getAvailablePort()
+	if err != nil {
+		return nil, fmt.Errorf("could not get a port: %v", err)
+	}
+	etcdURL = fmt.Sprintf("http://127.0.0.1:%d", etcdPort)
+	glog.Infof("starting etcd on %s", etcdURL)
+
+	etcdDataDir, err := ioutil.TempDir(os.TempDir(), "integration_test_etcd_data")
+	if err != nil {
+		return nil, fmt.Errorf("unable to make temp etcd data dir: %v", err)
+	}
+	glog.Infof("storing etcd data in: %v", etcdDataDir)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(
+		ctx,
+		etcdPath,
+		"--data-dir",
+		etcdDataDir,
+		"--listen-client-urls",
+		GetEtcdURL(),
+		"--advertise-client-urls",
+		GetEtcdURL(),
+		"--listen-peer-urls",
+		"http://127.0.0.1:0",
+	)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	stop := func() {
+		cancel()
+		err := cmd.Wait()
+		glog.Infof("etcd exit status: %v", err)
+		err = os.Remove(etcdDataDir)
 		if err != nil {
-			glog.Fatalf("Failed to run etcd: %v", err)
+			glog.Warningf("error during etcd cleanup: %v", err)
```

Comment:
I've tried running this locally and I'm seeing this constantly warning here with
```
W0901 00:06:50.067389      15 etcd.go:118] error during etcd cleanup: remove /tmp/integration_test_etcd_data551762727: directory not empty
```